### PR TITLE
Clean up asm

### PR DESCRIPTION
**Pull request by jeddenlea**
on an Intel(R) Xeon(R) CPU E5-2690 v2 @ 3.00GHz
```
on an Intel(R) Xeon(R) CPU E5-2690 v2 @ 3.00GHz
benchmark                       old ns/op     new ns/op     delta
BenchmarkBase32-40              6.05          5.71          -5.62%
BenchmarkLoop32-40              6.13          5.75          -6.20%
BenchmarkTail32_1-40            6.39          6.42          +0.47%
BenchmarkTail32_2-40            6.21          5.75          -7.41%
BenchmarkTail32_3-40            6.49          5.79          -10.79%
BenchmarkBasePartial32-40       113           116           +2.65%
BenchmarkLoopPartial32-40       114           119           +4.39%
BenchmarkTailPartial32_1-40     114           118           +3.51%
BenchmarkTailPartial32_2-40     117           120           +2.56%
BenchmarkTailPartial32_3-40     119           121           +1.68%
BenchmarkBase128-40             11.8          7.40          -37.29%
BenchmarkLoop128-40             13.2          9.91          -24.92%
BenchmarkTail128_1-40           11.8          8.42          -28.64%
BenchmarkTail128_2-40           11.8          8.80          -25.42%
BenchmarkTail128_3-40           11.8          9.14          -22.54%
BenchmarkTail128_4-40           11.5          9.51          -17.30%
BenchmarkTail128_5-40           11.8          10.2          -13.56%
BenchmarkTail128_6-40           11.5          10.2          -11.30%
BenchmarkTail128_7-40           11.5          10.8          -6.09%
BenchmarkTail128_8-40           11.7          10.5          -10.26%
BenchmarkTail128_9-40           12.6          11.2          -11.11%
BenchmarkTail128_10-40          12.9          11.8          -8.53%
BenchmarkTail128_11-40          12.9          12.2          -5.43%
BenchmarkTail128_12-40          13.0          12.2          -6.15%
BenchmarkTail128_13-40          13.3          13.1          -1.50%
BenchmarkTail128_14-40          13.5          13.2          -2.22%
BenchmarkTail128_15-40          13.6          13.2          -2.94%
```

### CODE REVIEW

**Comment by jeddenlea** [comment]
```
[jed@lappy]tmp$ benchcmp {old,new}_xeon
benchmark                old ns/op     new ns/op     delta
Benchmark32_1-40         5.83          6.40          +9.78%
Benchmark32_2-40         6.23          5.73          -8.03%
Benchmark32_4-40         6.15          5.39          -12.36%
Benchmark32_8-40         7.76          6.78          -12.63%
Benchmark32_16-40        10.5          9.28          -11.62%
Benchmark32_32-40        15.6          14.6          -6.41%
Benchmark32_64-40        27.7          26.7          -3.61%
Benchmark32_128-40       50.5          50.2          -0.59%
Benchmark32_256-40       110           107           -2.73%
Benchmark32_512-40       217           216           -0.46%
Benchmark32_1024-40      433           432           -0.23%
Benchmark32_2048-40      865           864           -0.12%
Benchmark32_4096-40      1727          1728          +0.06%
Benchmark32_8192-40      3453          3457          +0.12%
Benchmark128_1-40        11.8          8.44          -28.47%
Benchmark128_2-40        11.5          8.78          -23.65%
Benchmark128_4-40        11.8          9.46          -19.83%
Benchmark128_8-40        11.6          10.5          -9.48%
Benchmark128_16-40       13.5          9.95          -26.30%
Benchmark128_32-40       15.2          12.7          -16.45%
Benchmark128_64-40       21.0          19.2          -8.57%
Benchmark128_128-40      31.3          30.0          -4.15%
Benchmark128_256-40      53.6          52.8          -1.49%
Benchmark128_512-40      104           99.0          -4.81%
Benchmark128_1024-40     195           190           -2.56%
Benchmark128_2048-40     376           371           -1.33%
Benchmark128_4096-40     739           734           -0.68%
Benchmark128_8192-40     1463          1461          -0.14%

benchmark                old MB/s     new MB/s     speedup
Benchmark32_1-40         171.51       156.16       0.91x
Benchmark32_2-40         321.08       348.87       1.09x
Benchmark32_4-40         650.59       742.48       1.14x
Benchmark32_8-40         1030.63      1179.33      1.14x
Benchmark32_16-40        1524.10      1723.38      1.13x
Benchmark32_32-40        2049.80      2185.86      1.07x
Benchmark32_64-40        2313.08      2395.27      1.04x
Benchmark32_128-40       2533.36      2548.71      1.01x
Benchmark32_256-40       2323.06      2371.06      1.02x
Benchmark32_512-40       2349.99      2369.30      1.01x
Benchmark32_1024-40      2361.12      2369.38      1.00x
Benchmark32_2048-40      2367.23      2369.42      1.00x
Benchmark32_4096-40      2370.77      2369.14      1.00x
Benchmark32_8192-40      2372.26      2369.31      1.00x
Benchmark128_1-40        84.75        118.45       1.40x
Benchmark128_2-40        174.46       227.81       1.31x
Benchmark128_4-40        339.19       423.01       1.25x
Benchmark128_8-40        687.52       764.51       1.11x
Benchmark128_16-40       1186.63      1607.73      1.35x
Benchmark128_32-40       2109.00      2522.91      1.20x
Benchmark128_64-40       3043.83      3330.73      1.09x
Benchmark128_128-40      4083.94      4268.56      1.05x
Benchmark128_256-40      4774.09      4849.73      1.02x
Benchmark128_512-40      4901.69      5170.41      1.05x
Benchmark128_1024-40     5241.77      5379.17      1.03x
Benchmark128_2048-40     5439.65      5506.11      1.01x
Benchmark128_4096-40     5540.57      5574.32      1.01x
Benchmark128_8192-40     5596.12      5604.63      1.00x
```

**Comment by twmb** [comment]
On n1-standard-2 us-central1-b compute instance, 1.6.2:
```
benchmark               old ns/op     new ns/op     delta
Benchmark32_1-2         6.18          5.38          -12.94%
Benchmark32_2-2         6.47          5.44          -15.92%
Benchmark32_4-2         6.38          5.83          -8.62%
Benchmark32_8-2         7.68          7.41          -3.52%
Benchmark32_16-2        10.6          9.65          -8.96%
Benchmark32_22-2        12.3          12.0          -2.44%
Benchmark32_32-2        15.6          15.6          +0.00%
Benchmark32_64-2        28.8          28.8          +0.00%
Benchmark32_128-2       52.7          53.2          +0.95%
Benchmark32_256-2       117           117           +0.00%
Benchmark32_512-2       233           236           +1.29%
Benchmark32_1024-2      468           476           +1.71%
Benchmark32_2048-2      937           953           +1.71%
Benchmark32_4096-2      1873          1903          +1.60%
Benchmark32_8192-2      3757          3822          +1.73%
Benchmark32_16k-2       7521          7611          +1.20%
Benchmark32_32k-2       14980         15228         +1.66%
Benchmark128_1-2        10.4          8.93          -14.13%
Benchmark128_2-2        10.8          9.34          -13.52%
Benchmark128_4-2        11.1          9.71          -12.52%
Benchmark128_8-2        12.2          10.0          -18.03%
Benchmark128_16-2       11.8          10.2          -13.56%
Benchmark128_22-2       13.3          12.2          -8.27%
Benchmark128_32-2       14.0          13.0          -7.14%
Benchmark128_64-2       21.3          18.7          -12.21%
Benchmark128_128-2      33.3          31.5          -5.41%
Benchmark128_256-2      55.5          54.9          -1.08%
Benchmark128_512-2      102           104           +1.96%
Benchmark128_1024-2     201           199           -1.00%
Benchmark128_2048-2     391           389           -0.51%
Benchmark128_4096-2     776           769           -0.90%
Benchmark128_8192-2     1529          1530          +0.07%
Benchmark128_16k-2      3055          3057          +0.07%
Benchmark128_32k-2      6123          6149          +0.42%

benchmark               old MB/s     new MB/s     speedup
Benchmark32_1-2         161.77       185.80       1.15x
Benchmark32_2-2         309.12       367.38       1.19x
Benchmark32_4-2         626.92       685.68       1.09x
Benchmark32_8-2         1041.08      1079.62      1.04x
Benchmark32_16-2        1502.35      1658.83      1.10x
Benchmark32_22-2        1788.02      1829.25      1.02x
Benchmark32_32-2        2044.89      2054.11      1.00x
Benchmark32_64-2        2218.88      2224.12      1.00x
Benchmark32_128-2       2431.06      2404.23      0.99x
Benchmark32_256-2       2183.15      2181.80      1.00x
Benchmark32_512-2       2191.57      2162.69      0.99x
Benchmark32_1024-2      2187.56      2148.05      0.98x
Benchmark32_2048-2      2183.51      2148.76      0.98x
Benchmark32_4096-2      2186.01      2152.00      0.98x
Benchmark32_8192-2      2180.32      2143.16      0.98x
Benchmark32_16k-2       2178.18      2152.66      0.99x
Benchmark32_32k-2       2187.45      2151.80      0.98x
Benchmark128_1-2        95.91        111.99       1.17x
Benchmark128_2-2        184.70       214.09       1.16x
Benchmark128_4-2        361.29       412.01       1.14x
Benchmark128_8-2        653.28       797.27       1.22x
Benchmark128_16-2       1354.09      1573.96      1.16x
Benchmark128_22-2       1657.03      1804.62      1.09x
Benchmark128_32-2       2291.89      2456.17      1.07x
Benchmark128_64-2       3002.45      3419.30      1.14x
Benchmark128_128-2      3848.38      4069.82      1.06x
Benchmark128_256-2      4615.59      4664.45      1.01x
Benchmark128_512-2      4979.39      4882.39      0.98x
Benchmark128_1024-2     5085.26      5121.31      1.01x
Benchmark128_2048-2     5228.19      5256.75      1.01x
Benchmark128_4096-2     5273.05      5321.84      1.01x
Benchmark128_8192-2     5356.31      5353.50      1.00x
Benchmark128_16k-2      5361.54      5358.48      1.00x
Benchmark128_32k-2      5350.79      5328.86      1.00x
```

same instance, tip: (`go version devel +85e39f8 Fri May 20 17:10:18 2016 +0000 linux/amd64`):
```
benchmark               old ns/op     new ns/op     delta
Benchmark32_1-2         5.97          5.30          -11.22%
Benchmark32_2-2         6.28          5.45          -13.22%
Benchmark32_4-2         6.15          5.58          -9.27%
Benchmark32_8-2         7.61          6.80          -10.64%
Benchmark32_16-2        10.3          9.63          -6.50%
Benchmark32_22-2        12.1          11.9          -1.65%
Benchmark32_32-2        15.7          15.2          -3.18%
Benchmark32_64-2        28.8          29.1          +1.04%
Benchmark32_128-2       59.3          52.7          -11.13%
Benchmark32_256-2       117           115           -1.71%
Benchmark32_512-2       235           236           +0.43%
Benchmark32_1024-2      470           470           +0.00%
Benchmark32_2048-2      933           944           +1.18%
Benchmark32_4096-2      1877          1875          -0.11%
Benchmark32_8192-2      3750          3781          +0.83%
Benchmark32_16k-2       7590          7603          +0.17%
Benchmark32_32k-2       15188         15188         +0.00%
Benchmark128_1-2        10.4          9.30          -10.58%
Benchmark128_2-2        10.7          9.65          -9.81%
Benchmark128_4-2        11.1          10.1          -9.01%
Benchmark128_8-2        12.2          10.4          -14.75%
Benchmark128_16-2       12.3          10.0          -18.70%
Benchmark128_22-2       13.5          12.3          -8.89%
Benchmark128_32-2       14.2          12.7          -10.56%
Benchmark128_64-2       21.0          18.4          -12.38%
Benchmark128_128-2      32.9          31.2          -5.17%
Benchmark128_256-2      55.1          54.5          -1.09%
Benchmark128_512-2      103           102           -0.97%
Benchmark128_1024-2     202           198           -1.98%
Benchmark128_2048-2     391           391           +0.00%
Benchmark128_4096-2     776           765           -1.42%
Benchmark128_8192-2     1544          1527          -1.10%
Benchmark128_16k-2      3083          3042          -1.33%
Benchmark128_32k-2      6068          6097          +0.48%

benchmark               old MB/s     new MB/s     speedup
Benchmark32_1-2         167.49       188.56       1.13x
Benchmark32_2-2         318.49       367.26       1.15x
Benchmark32_4-2         650.48       716.27       1.10x
Benchmark32_8-2         1051.20      1175.70      1.12x
Benchmark32_16-2        1556.90      1661.55      1.07x
Benchmark32_22-2        1814.90      1850.16      1.02x
Benchmark32_32-2        2043.36      2101.63      1.03x
Benchmark32_64-2        2222.83      2198.38      0.99x
Benchmark32_128-2       2159.35      2426.97      1.12x
Benchmark32_256-2       2177.89      2220.65      1.02x
Benchmark32_512-2       2171.84      2166.90      1.00x
Benchmark32_1024-2      2175.37      2175.00      1.00x
Benchmark32_2048-2      2194.89      2168.47      0.99x
Benchmark32_4096-2      2181.72      2183.53      1.00x
Benchmark32_8192-2      2184.08      2166.15      0.99x
Benchmark32_16k-2       2158.53      2154.76      1.00x
Benchmark32_32k-2       2157.38      2157.45      1.00x
Benchmark128_1-2        96.46        107.58       1.12x
Benchmark128_2-2        186.79       207.27       1.11x
Benchmark128_4-2        360.70       397.63       1.10x
Benchmark128_8-2        657.37       766.56       1.17x
Benchmark128_16-2       1301.72      1594.44      1.22x
Benchmark128_22-2       1633.83      1794.39      1.10x
Benchmark128_32-2       2250.64      2514.41      1.12x
Benchmark128_64-2       3048.43      3469.59      1.14x
Benchmark128_128-2      3885.10      4097.30      1.05x
Benchmark128_256-2      4643.92      4694.06      1.01x
Benchmark128_512-2      4962.42      5017.08      1.01x
Benchmark128_1024-2     5053.39      5153.70      1.02x
Benchmark128_2048-2     5228.93      5235.24      1.00x
Benchmark128_4096-2     5277.55      5350.58      1.01x
Benchmark128_8192-2     5302.86      5361.35      1.01x
Benchmark128_16k-2      5313.45      5385.51      1.01x
Benchmark128_32k-2      5399.74      5374.36      1.00x
```

Intel 32 core @ CPU2899.660MHz, tip still:
```
benchmark                old ns/op     new ns/op     delta
Benchmark32_1-32         6.97          6.87          -1.43%
Benchmark32_2-32         7.47          6.30          -15.66%
Benchmark32_4-32         7.74          6.67          -13.82%
Benchmark32_8-32         9.38          8.43          -10.13%
Benchmark32_16-32        13.2          11.8          -10.61%
Benchmark32_22-32        14.8          14.4          -2.70%
Benchmark32_32-32        19.0          19.0          +0.00%
Benchmark32_64-32        33.7          33.1          -1.78%
Benchmark32_128-32       65.9          61.9          -6.07%
Benchmark32_256-32       125           121           -3.20%
Benchmark32_512-32       245           238           -2.86%
Benchmark32_1024-32      478           475           -0.63%
Benchmark32_2048-32      938           945           +0.75%
Benchmark32_4096-32      1892          1921          +1.53%
Benchmark32_8192-32      3742          3822          +2.14%
Benchmark32_16k-32       7451          7579          +1.72%
Benchmark32_32k-32       14981         15189         +1.39%
Benchmark128_1-32        12.9          10.4          -19.38%
Benchmark128_2-32        13.3          10.5          -21.05%
Benchmark128_4-32        13.5          11.1          -17.78%
Benchmark128_8-32        13.8          11.8          -14.49%
Benchmark128_16-32       14.9          12.0          -19.46%
Benchmark128_22-32       16.4          14.6          -10.98%
Benchmark128_32-32       17.8          15.4          -13.48%
Benchmark128_64-32       23.3          21.9          -6.01%
Benchmark128_128-32      35.9          35.4          -1.39%
Benchmark128_256-32      62.0          61.7          -0.48%
Benchmark128_512-32      114           115           +0.88%
Benchmark128_1024-32     223           219           -1.79%
Benchmark128_2048-32     434           428           -1.38%
Benchmark128_4096-32     849           843           -0.71%
Benchmark128_8192-32     1688          1706          +1.07%
Benchmark128_16k-32      3360          3401          +1.22%
Benchmark128_32k-32      6690          6794          +1.55%

benchmark                old MB/s     new MB/s     speedup
Benchmark32_1-32         143.41       145.54       1.01x
Benchmark32_2-32         267.80       317.46       1.19x
Benchmark32_4-32         517.03       599.57       1.16x
Benchmark32_8-32         852.84       948.92       1.11x
Benchmark32_16-32        1214.03      1356.52      1.12x
Benchmark32_22-32        1483.75      1530.79      1.03x
Benchmark32_32-32        1680.35      1681.65      1.00x
Benchmark32_64-32        1897.88      1933.61      1.02x
Benchmark32_128-32       1942.14      2068.34      1.06x
Benchmark32_256-32       2035.65      2105.69      1.03x
Benchmark32_512-32       2088.90      2150.75      1.03x
Benchmark32_1024-32      2139.77      2153.19      1.01x
Benchmark32_2048-32      2181.50      2166.72      0.99x
Benchmark32_4096-32      2164.43      2131.64      0.98x
Benchmark32_8192-32      2189.19      2143.28      0.98x
Benchmark32_16k-32       2198.88      2161.70      0.98x
Benchmark32_32k-32       2187.28      2157.31      0.99x
Benchmark128_1-32        77.41        96.49        1.25x
Benchmark128_2-32        150.28       191.26       1.27x
Benchmark128_4-32        296.67       361.54       1.22x
Benchmark128_8-32        578.18       679.81       1.18x
Benchmark128_16-32       1073.54      1338.16      1.25x
Benchmark128_22-32       1342.88      1502.13      1.12x
Benchmark128_32-32       1796.93      2075.40      1.15x
Benchmark128_64-32       2741.23      2916.60      1.06x
Benchmark128_128-32      3566.57      3610.72      1.01x
Benchmark128_256-32      4129.27      4146.72      1.00x
Benchmark128_512-32      4470.79      4446.87      0.99x
Benchmark128_1024-32     4575.14      4661.00      1.02x
Benchmark128_2048-32     4713.71      4781.43      1.01x
Benchmark128_4096-32     4822.27      4854.51      1.01x
Benchmark128_8192-32     4851.12      4800.75      0.99x
Benchmark128_16k-32      4875.35      4817.20      0.99x
Benchmark128_32k-32      4897.98      4822.98      0.98x
```

I'll review this weekend. Just wanted to grab a few more benchmarks specifically against tip.

File: murmur32_amd64.s [hunk 1]
```diff
@@ -1,100 +1,92 @@
...
 #define c1_32 0xcc9e2d51
 #define c2_32 0x1b873593
 
-// AX = h1
-// SI = &data[0]
+// Expects:
+// R12 == uint32 seed
+// SI  == &data
+// R9  == len(data)
+// BX  == &[2]uint64 return
```
Comment (twmb) [inline murmur32_amd64.s]: &uint32

File: murmur32_amd64.s [hunk 2]
```diff
@@ -1,100 +1,92 @@
...
+	JZ tail2
+
+tail3: // no jump
+	MOVB 2(SI)(BP*1), AX
+	SALL $16, AX
+tail2:
+	MOVW (SI)(BP*1), AX
+	JMP fintail
+tail1:
+	MOVB (SI)(BP*1), AL
```
Comment (twmb) [inline murmur32_amd64.s]: Why `AL` here as opposed to `AX` for consistency?

**Comment by twmb** [comment]
For posterity, the diff of working from -nblocks to 0 for the loop:

```
diff --git a/murmur32_amd64.s b/murmur32_amd64.s
index 6e9a4bb..c495b1f 100644
--- a/murmur32_amd64.s
+++ b/murmur32_amd64.s
@@ -25,15 +25,17 @@ TEXT ·SeedSum32(SB), $0-36
 // R9  == len(data)
 // BX  == &[2]uint64 return
 TEXT sum32internal<>(SB), $0
-	XORQ BP, BP
 	MOVQ R9, CX
-	ANDQ $-4, CX // cx = len(data) - (len(data) % 4)
+	SHRQ $2, CX
 	JZ   tail

-	// for bp = 0 ; bp < cx; bp += 4 {...
+	MOVQ CX, BP
+	LEAQ (SI)(BP*4), SI
+	NEGQ BP
+
+	// for bp = -blocks ; bp != 0; bp++ {...
 loop:
-	MOVL (SI)(BP*1), AX
-	ADDQ $4, BP
+	MOVL (SI)(BP*4), AX
+      ADDQ $1, BP

 	IMULL $c1_32, AX
 	ROLL  $15, AX
@@ -43,8 +45,8 @@ loop:
 	ROLL $13, R12
 	LEAL 0xe6546b64(R12)(R12*4), R12

-	CMPQ BP, CX
-	JNE  loop
+	TEST BP, BP
+	JNZ  loop

 tail:
 	MOVQ R9, CX
@@ -59,15 +61,15 @@ tail:
 	JZ   tail2

 tail3:  // no jump
-	MOVB 2(SI)(BP*1), AX
+	MOVB 2(SI), AX
 	SALL $16, AX

 tail2:
-	MOVW (SI)(BP*1), AX
+	MOVW (SI), AX
 	JMP  fintail

 tail1:
-	MOVB (SI)(BP*1), AL
+	MOVB (SI), AL

 fintail:
 	IMULL $c1_32, AX
```

actually slows things down.

File: murmur128_amd64.s [hunk 3]
```diff
@@ -1,256 +1,190 @@
...
+tail11:
+	MOVBQZX 10(SI)(BP*1), AX
+	SALQ $16, AX
+tail10:
+	MOVW 8(SI)(BP*1), AX
+	JMP fintailhigh
+tail9:
+	MOVB 8(SI)(BP*1), AL
+fintailhigh:
+	MULQ R15
```
Comment (twmb) [inline murmur128_amd64.s]: Does having the implicit `AX` save time?

File: murmur32_amd64.s [hunk 4]
```diff
@@ -1,100 +1,92 @@
...
+// Expects:
+// R12 == uint32 seed
+// SI  == &data
+// R9  == len(data)
+// BX  == &[2]uint64 return
+TEXT sum32internal<>(SB), $0
+	XORQ BP, BP
+	MOVQ R9, CX
+	ANDQ $-4, CX // cx = len(data) - (len(data) % 4)
+	JZ tail
```
Comment (twmb) [inline murmur32_amd64.s]: In `sum128internal` you do not have this optimistic `JZ tail`, but instead have `CMP ... ; JZ ... ` at the beginning of `loop` and an unconditional `JMP loop` at the end. It'd be better to be consistent (and one approach is probably better). Might be worth benchmarking.

**Comment by twmb** [comment]
Mostly LGTM; I have some questions. This could also go through [`asmfmt`](https://github.com/klauspost/asmfmt).

**Comment by twmb** [comment]
Running benchmarks against PR #5 on the 32 core server demonstrates the strengths of these tail optimizations even better:

```
benchmark                       old ns/op     new ns/op     delta
BenchmarkBase32-32              6.32          6.06          -4.11%
BenchmarkLoop32-32              7.65          6.71          -12.29%
BenchmarkTail32_1-32            7.09          6.68          -5.78%
BenchmarkTail32_2-32            7.37          6.43          -12.75%
BenchmarkTail32_3-32            8.06          6.69          -17.00%
BenchmarkBasePartial32-32       112           112           +0.00%
BenchmarkLoopPartial32-32       114           114           +0.00%
BenchmarkTailPartial32_1-32     114           114           +0.00%
BenchmarkTailPartial32_2-32     115           115           +0.00%
BenchmarkTailPartial32_3-32     119           118           -0.84%
BenchmarkBase128-32             13.9          8.44          -39.28%
BenchmarkLoop128-32             15.5          11.9          -23.23%
BenchmarkTail128_1-32           14.2          10.0          -29.58%
BenchmarkTail128_2-32           14.2          10.3          -27.46%
BenchmarkTail128_3-32           13.8          10.7          -22.46%
BenchmarkTail128_4-32           14.0          11.2          -20.00%
BenchmarkTail128_5-32           14.0          12.0          -14.29%
BenchmarkTail128_6-32           14.0          11.7          -16.43%
BenchmarkTail128_7-32           13.8          12.3          -10.87%
BenchmarkTail128_8-32           13.8          12.2          -11.59%
BenchmarkTail128_9-32           15.0          13.6          -9.33%
BenchmarkTail128_10-32          15.3          13.6          -11.11%
BenchmarkTail128_11-32          15.5          14.3          -7.74%
BenchmarkTail128_12-32          15.9          14.3          -10.06%
BenchmarkTail123_13-32          16.0          15.0          -6.25%
BenchmarkTail128_14-32          16.2          16.1          -0.62%
BenchmarkTail128_15-32          16.4          15.6          -4.88%

benchmark                       old MB/s     new MB/s     speedup
BenchmarkLoop32-32              522.59       596.43       1.14x
BenchmarkTail32_1-32            141.10       149.62       1.06x
BenchmarkTail32_2-32            271.41       310.86       1.15x
BenchmarkTail32_3-32            372.41       448.25       1.20x
BenchmarkLoopPartial32-32       34.81        34.87        1.00x
BenchmarkTailPartial32_1-32     8.72         8.74         1.00x
BenchmarkTailPartial32_2-32     17.38        17.33        1.00x
BenchmarkTailPartial32_3-32     25.13        25.31        1.01x
BenchmarkLoop128-32             1032.35      1348.84      1.31x
BenchmarkTail128_1-32           70.59        99.62        1.41x
BenchmarkTail128_2-32           140.61       193.89       1.38x
BenchmarkTail128_3-32           217.17       281.36       1.30x
BenchmarkTail128_4-32           285.59       358.74       1.26x
BenchmarkTail128_5-32           357.00       418.05       1.17x
BenchmarkTail128_6-32           428.72       513.69       1.20x
BenchmarkTail128_7-32           505.43       569.13       1.13x
BenchmarkTail128_8-32           578.98       656.32       1.13x
BenchmarkTail128_9-32           598.69       661.79       1.11x
BenchmarkTail128_10-32          651.76       734.21       1.13x
BenchmarkTail128_11-32          707.49       770.08       1.09x
BenchmarkTail128_12-32          754.05       839.56       1.11x
BenchmarkTail123_13-32          814.54       865.33       1.06x
BenchmarkTail128_14-32          865.42       869.05       1.00x
BenchmarkTail128_15-32          914.61       962.90       1.05x
```

File: murmur128_amd64.s (hunk 3 above)
Comment (twmb) [inline murmur128_amd64.s]: For fun, I changed these four `MULQ` to `IMULQ` with `, AX`:

```
BenchmarkTail128_1-32           10.0          9.80          -2.00%
BenchmarkTail128_2-32           10.3          9.74          -5.44%
BenchmarkTail128_3-32           10.7          10.3          -3.74%
BenchmarkTail128_4-32           11.2          10.3          -8.04%
BenchmarkTail128_5-32           12.0          11.3          -5.83%
BenchmarkTail128_6-32           11.7          11.4          -2.56%
BenchmarkTail128_7-32           12.3          12.3          +0.00%
BenchmarkTail128_8-32           12.2          11.6          -4.92%
BenchmarkTail128_9-32           13.6          12.7          -6.62%
BenchmarkTail128_10-32          13.6          13.2          -2.94%
BenchmarkTail128_11-32          14.3          13.4          -6.29%
BenchmarkTail128_12-32          14.3          13.3          -6.99%
BenchmarkTail123_13-32          15.0          14.1          -6.00%
BenchmarkTail128_14-32          16.1          14.4          -10.56%
BenchmarkTail128_15-32          15.6          14.4          -7.69%
```

Probably worth it to change these to `IMULQ %reg, AX`

File: murmur32_amd64.s (hunk 4 above)
Comment (twmb) [inline murmur32_amd64.s]: It seems negligible and oddly, in my benchmarks JZ here but no JZ in 128 seemed to be the better options.

File: murmur32_amd64.s (hunk 2 above)
Comment (jeddenlea) [inline murmur32_amd64.s]: With a byte operation, either AL or AH can be used.  AL explicitly states which byte of AX I want to set.

File: murmur32_amd64.s (hunk 1 above)
Comment (jeddenlea) [inline murmur32_amd64.s]: derp

**Pull request merged**